### PR TITLE
Add interval-partitioner as a pass

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

Adds the interval partitioner used in the cuda code-generation as a pass that can be enabled. 
Also adds the pretty-printer of the stencil-instantiation

#### Resolves / Enhances

This is used in cuda code-gen as well as helps to write DACE-sdfg's

## Dependencies

This PR is independent

### CODE REVIEW

**Comment by twicki** [comment]
launch pipeline

File: src/dawn/Optimizer/PassIntervalPartitioner.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+///
+/// This pass is not necessary to create legal code and is hence not in the debug-group
+class PassIntervalPartitioner : public Pass {
+public:
+  PassIntervalPartitioner(OptimizerContext& context) : Pass(context, "PassIntervalPartitioner") {}
+
+  /// @brief Pass implementation
+  bool run(const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation) override;
+
+private:
```
Comment (lukasm91) [inline src/dawn/Optimizer/PassIntervalPartitioner.h]:
```suggestion
```

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿//===--------------------------------------------------------------------------------*- C++ -*-===//
+//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (lukasm91) [inline src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: what happened here?

File: src/dawn/Optimizer/PassIntervalPartitioner.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
+﻿//===--------------------------------------------------------------------------------*- C++ -*-===//
```
Comment (lukasm91) [inline src/dawn/Optimizer/PassIntervalPartitioner.cpp]: You will need to explain me the idea of this pass.

File: src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp (hunk 2 above)
Comment (twicki) [reply src/dawn/Optimizer/PassTemporaryToStencilFunction.cpp]: I can't say

File: src/dawn/Optimizer/PassIntervalPartitioner.cpp (hunk 3 above)
Comment (twicki) [reply src/dawn/Optimizer/PassIntervalPartitioner.cpp]: Assume we have a multistage with intervals [start : end], [start+1 : end-1] [start : start+4] [end-2 : end]

The pass will split those into disjoint intervals:

[start : start], [start+1 : start+4], [start+5 : end-3], [end-2 : end-1], [end : end] and will add multiple do-methods with these intervals for each computation

File: src/dawn/Optimizer/PassIntervalPartitioner.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      auto cloneMS = multiStage->clone();
+
+      auto multiInterval = multiStage->computePartitionOfIntervals();
+
+      // empty out the multistage
+      for(auto cloneStageIt = cloneMS->childrenBegin(); cloneStageIt != cloneMS->childrenEnd();
+          ++cloneStageIt) {
+        auto cloneDoMethodIt = (*cloneStageIt)->childrenBegin();
+        while(cloneDoMethodIt != (*cloneStageIt)->childrenEnd()) {
+          cloneDoMethodIt = (*cloneStageIt)->childrenErase(cloneDoMethodIt);
```
Comment (lukasm91) [inline src/dawn/Optimizer/PassIntervalPartitioner.cpp]: instead of cloning the MS and deleting its data, can't we just create a new multistage?

File: src/dawn/Optimizer/StatementMapper.cpp [hunk 5]
```diff
@@ -120,37 +120,42 @@ void StatementMapper::visit(const std::shared_ptr<iir::IfStmt>& stmt) {
 void StatementMapper::visit(const std::shared_ptr<iir::VarDeclStmt>& stmt) {
   DAWN_ASSERT(initializedWithBlockStmt_);
 
-  // This is the first time we encounter this variable. We have to make sure the name is not
-  // already used in another scope!
-  int AccessID = instantiation_->nextUID();
+  int accessID = -1;
+  if(metadata_.hasStmtToAccessID(stmt)) {
```
Comment (lukasm91) [inline src/dawn/Optimizer/StatementMapper.cpp]: is this change needed because after duplicating a do-method, we might have two local variables with the same name?
